### PR TITLE
[refactor](queryqueue) using a priority queue in query queue in order to implement priority management in the future

### PR DESCRIPTION
**Pull request by yiguolei**
## Proposed changes

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/resource/workloadgroup/QueueOfferToken.java [hunk 1]
```diff
@@ -17,30 +17,101 @@
...
 package org.apache.doris.resource.workloadgroup;
 
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.ReentrantLock;
+
 // used to mark QueryQueue offer result
 // if offer failed, then need to cancel query
 // and return failed reason to user client
 public class QueueOfferToken {
```
Comment (wangbo) [inline fe/fe-core/src/main/java/org/apache/doris/resource/workloadgroup/QueueOfferToken.java:27]:
```suggestion
public class QueueToken {
```

File: fe/fe-core/src/main/java/org/apache/doris/resource/workloadgroup/QueueOfferToken.java [hunk 2]
```diff
@@ -17,30 +17,101 @@
...
+
+    static AtomicLong tokenIdGenerator = new AtomicLong(0);
+
+    private long tokenId = 0;
+
+    private TokenState tokenState;
+
+    private long waitTimeout = 0;
+
+    private long enqueueTime = 0;
```
Comment (wangbo) [inline fe/fe-core/src/main/java/org/apache/doris/resource/workloadgroup/QueueOfferToken.java:43]: unused now, drop it.

Comment (yiguolei) [reply fe/fe-core/src/main/java/org/apache/doris/resource/workloadgroup/QueueOfferToken.java:43]: No, I will use it in the future. I will display enqueue time in wg tvf or show process list.

**Review by xinyiZzz — DISMISSED** [review]
LGTM

**Comment by yiguolei** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/qe/StmtExecutor.java [hunk 3]
```diff
@@ -607,23 +607,29 @@ public void finalizeQuery() {
...
                 throw new RuntimeException("interrupted Exception happens when queue query");
             }
-            if (offerRet != null && !offerRet.isOfferSuccess()) {
+            if (offerRet != null && !offerRet.enqueueSuccess()) {
                 String retMsg = "queue failed, reason=" + offerRet.getOfferResultDetail();
                 LOG.error("query (id=" + DebugUtil.printId(queryId) + ") " + retMsg);
                 throw new UserException(retMsg);
             }
+            if (!offerRet.waitSignal()) {
+                String retMsg = "queue success but wait too long in queue";
```
Comment (wangbo) [inline fe/fe-core/src/main/java/org/apache/doris/qe/StmtExecutor.java:628]:
```suggestion
                String retMsg = offerRet.getOfferResultDetail();
```

File: fe/fe-core/src/main/java/org/apache/doris/qe/StmtExecutor.java [hunk 4]
```diff
@@ -607,23 +607,29 @@ public void finalizeQuery() {
     private void handleQueryWithRetry(TUniqueId queryId) throws Exception {
         // queue query here
         syncJournalIfNeeded();
-        QueueOfferToken offerRet = null;
+        QueueToken offerRet = null;
```
Comment (wangbo) [inline fe/fe-core/src/main/java/org/apache/doris/qe/StmtExecutor.java:610]:
```suggestion
        QueueToken queueToken = null;
```

File: fe/fe-core/src/main/java/org/apache/doris/resource/workloadgroup/QueryQueue.java [hunk 5]
```diff
@@ -91,39 +94,32 @@ public String debugString() {
                 + ", currentWaitingQueryNum=" + currentWaitingQueryNum;
     }
 
-    public QueueOfferToken offer() throws InterruptedException {
+    public QueueToken getToken() throws InterruptedException {
         // to prevent hang
         // the lock shouldn't be hold for too long
         // we should catch the case when it happens
```
Comment (wangbo) [inline fe/fe-core/src/main/java/org/apache/doris/resource/workloadgroup/QueryQueue.java:100]: remove the comment.

File: fe/fe-core/src/main/java/org/apache/doris/resource/workloadgroup/QueryQueue.java [hunk 6]
```diff
@@ -132,14 +128,29 @@ public QueueOfferToken offer() throws InterruptedException {
...
+            }
             currentRunningQueryNum--;
             Preconditions.checkArgument(currentRunningQueryNum >= 0);
             // maybe only when currentWaitingQueryNum != 0 need to signal
             if (currentRunningQueryNum < maxConcurrency) {
-                queueLockCond.signal();
+                QueueToken nextToken = this.priorityTokenQueue.poll();
+                if (nextToken != null) {
+                    ++currentRunningQueryNum;
+                    nextToken.signal();
```
Comment (wangbo) [inline fe/fe-core/src/main/java/org/apache/doris/resource/workloadgroup/QueryQueue.java:152]: Add log for signal and wait Exception

File: fe/fe-core/src/main/java/org/apache/doris/resource/workloadgroup/QueueToken.java [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+    private final ReentrantLock tokenLock = new ReentrantLock();
+    private final Condition tokenCond = tokenLock.newCondition();
+
+    public QueueToken(TokenState tokenState, long waitTimeout,
+            String offerResultDetail) {
+        this.tokenId = tokenIdGenerator.addAndGet(1);
+        this.tokenState = tokenState;
+        this.waitTimeout = waitTimeout;
+        this.offerResultDetail = offerResultDetail;
+        this.enqueueTime = System.currentTimeMillis();
```
Comment (wangbo) [inline fe/fe-core/src/main/java/org/apache/doris/resource/workloadgroup/QueueToken.java:56]: enqueueTime is useless, remove it.

File: fe/fe-core/src/main/java/org/apache/doris/resource/workloadgroup/QueryQueue.java [hunk 8]
```diff
@@ -132,14 +128,29 @@ public QueueOfferToken offer() throws InterruptedException {
         }
     }
 
-    public void poll() throws InterruptedException {
-        queueLock.tryLock(5, TimeUnit.SECONDS);
+    // If the token is acquired and do work success, then call this method to release it.
+    public void returnToken(QueueToken token) throws InterruptedException {
```
Comment (wangbo) [inline fe/fe-core/src/main/java/org/apache/doris/resource/workloadgroup/QueryQueue.java:132]:
```suggestion
    public void dequeueToken(QueueToken token) throws InterruptedException {
```

Comment (yiguolei) [reply fe/fe-core/src/main/java/org/apache/doris/resource/workloadgroup/QueryQueue.java:132]: 是不是不是dequeue，dequeue 是出队列的意思，我们是归还token

**Review by wangbo — DISMISSED** [review]
LGTM

**Comment by yiguolei** [comment]
run buildall

**Comment by yiguolei** [comment]
run buildall

**Review by wangbo — DISMISSED** [review]
LGTM

**Comment by yiguolei** [comment]
run buildall

**Comment by yiguolei** [comment]
run buildall

**Comment by yiguolei** [comment]
run feut

**Review by xinyiZzz — DISMISSED** [review]
LGTM

File: fe/fe-core/src/main/java/org/apache/doris/qe/StmtExecutor.java [hunk 9]
```diff
@@ -607,23 +607,28 @@ public void finalizeQuery() {
...
                 throw new RuntimeException("interrupted Exception happens when queue query");
             }
-            if (offerRet != null && !offerRet.isOfferSuccess()) {
-                String retMsg = "queue failed, reason=" + offerRet.getOfferResultDetail();
+            if (queueToken != null && !queueToken.enqueueSuccess()) {
+                String retMsg = "queue failed, reason=" + queueToken.getOfferResultDetail();
                 LOG.error("query (id=" + DebugUtil.printId(queryId) + ") " + retMsg);
                 throw new UserException(retMsg);
             }
+            if (!queueToken.waitSignal()) {
```
Comment (wangbo) [inline fe/fe-core/src/main/java/org/apache/doris/qe/StmtExecutor.java:627]: if (queueToken != null && !queueToken.waitSignal())

Comment (wangbo) [inline fe/fe-core/src/main/java/org/apache/doris/qe/StmtExecutor.java:627]:
```suggestion
            if (queueToken!= null && !queueToken.waitSignal()) {
```

Comment (yiguolei) [reply fe/fe-core/src/main/java/org/apache/doris/qe/StmtExecutor.java:627]: It will never be null 啊

**Comment by yiguolei** [comment]
run buildall

**Comment by yiguolei** [comment]
run buildall

**Review by xinyiZzz — APPROVED** [review]
LGTM

**Review by wangbo — APPROVED** [review]
LGTM

**Pull request merged**